src/HLSLTree.cpp

```diff
@@ -382,6 +382,40 @@ bool HLSLTree::GetExpressionValue(HLSLExpression * expression, int & value)
     return false;
 }
 
+bool HLSLTree::NeedsFunction(const char* name)
+{
+    // Early out
+    if (!GetContainsString(name))
+        return false;
+
+    struct NeedsFunctionVisitor: HLSLTreeVisitor
```

Comment:
NeedsFunction is completely invisible on the profile (in fact, MSL generation is ~10x faster than HLSL parsing on our shaders - and as mentioned before the actual Metal/clang parsing & generation is much slower than that).